app/experimenter/nimbus-ui/src/components/ChangeApprovalOperations/index.tsx

```diff
@@ -49,30 +57,55 @@ export const ChangeApprovalOperations: React.FC<
   startRemoteSettingsApproval,
   children,
 }) => {
-  let defaultUIState = ChangeApprovalOperationsState.None;
-  if (rejectionEvent) {
-    defaultUIState = ChangeApprovalOperationsState.Rejected;
-  } else if (timeoutEvent) {
-    defaultUIState = canReview
-      ? ChangeApprovalOperationsState.ApproveOrReject
-      : ChangeApprovalOperationsState.ApprovalPending;
-  } else if (approvalEvent) {
-    defaultUIState = canReview
-      ? ChangeApprovalOperationsState.RemoteSettingsPending
-      : ChangeApprovalOperationsState.ApprovalPending;
-  } else if (reviewRequestEvent) {
-    defaultUIState = canReview
-      ? ChangeApprovalOperationsState.ApproveOrReject
-      : ChangeApprovalOperationsState.ApprovalPending;
-  }
+  const uiStateDependencies = [
+    canReview,
+    reviewRequestEvent,
+    approvalEvent,
+    rejectionEvent,
+    timeoutEvent,
+  ];
+
+  const defaultUIState = useMemo(() => {
+    switch (
+      determineLatestChangelogEventKind({
+        reviewRequestEvent,
+        approvalEvent,
+        rejectionEvent,
+        timeoutEvent,
+      })
+    ) {
+      case EventKinds.Rejection:
+        return ChangeApprovalOperationsState.Rejected;
+      case EventKinds.Approval:
+        return canReview
+          ? ChangeApprovalOperationsState.RemoteSettingsPending
+          : ChangeApprovalOperationsState.ApprovalPending;
+      case EventKinds.Request:
+        return canReview
+          ? ChangeApprovalOperationsState.ApproveOrReject
+          : ChangeApprovalOperationsState.ApprovalPending;
+      case EventKinds.Timeout:
+        return canReview
+          ? ChangeApprovalOperationsState.ApproveOrReject
+          : ChangeApprovalOperationsState.ApprovalPending;
+      default:
+        return ChangeApprovalOperationsState.None;
+    }
+  }, uiStateDependencies);
```

Comment:
This ended up expanding in complexity a bit. Best solution I could come up with is to find which is the most recent of all the significant changelog events and treat that as the current state. That seems to correctly handle the case of a retry request after a rejection and a retry approval after a timeout.

Reply:
Shouldn't you be able to switch this on the pair of status/publish_status rather than looking at the most recent changelog?

Reply:
Oh yeah that makes sense, it should show the current UI state based solely on the publish_status, the status should just determine whether it's launching or ending.  I'm hesitant about building logic strongly around changelogs, since they can change and move around a lot more than the model does.